File: conserved_resides/conserve.cpp
//
//  conserve.cpp
//  h+states
//
//  Created by Rohith Pentaparthy on 8/2/17.
//  Copyright © 2017 Rohith. All rights reserved.
//

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>
using namespace std;
void get_conserved_residues(vector<int> &hxt36, vector<int> &xyle);

void get_conserved_residues(vector<int> &hxt36, vector<int> &xyle) {
	// hxt36 input
	freopen("/Users/rohith/Documents/MayesLab/h+states/color_data_hxt36.txt", "r", stdin);
	double input = 0;
	int res_num = 49;

// VERY BUGGY, I'm purposely reading in 1 less than the total number in
// the hxt36 file, because if I read the last number, then all of the
// xyle values become 0's for some reason
// I also can't make input an int...
	for (int i = 0; i < 483; ++i) {
		cin >> input;
		//cout << input << " ";
		if (input == 1) hxt36.push_back(res_num);
		res_num++;
	}
	//cout << "\n";
	fclose(stdin);
	
	// xyle input
	freopen("/Users/rohith/Documents/MayesLab/h+states/color_data_xyle.txt", "r", stdin);
	res_num = 5;
	for (int i = 0; i < 475; ++i) {
		cin >> input;
		//cout << input << " ";
		if (input == 1) xyle.push_back(res_num);
		res_num++;
	}
	//cout << "\n";
	fclose(stdin);
	return;
}



/*** MAIN ***/
int main() {
	// insert code here...
	// freopen needs to be commented out or pathway needs to be modified
	
	
	
	vector<int> hxt36;
	vector<int> xyle;
	get_conserved_residues(hxt36, xyle);
	
//	cout << "hxt36 conserved residues:\n";
//	for (size_t i = 0; i < hxt36.size(); ++i) {
//		cout << hxt36[i] << '\n';
//	}
//	cout << "\n\n";
//	
//	cout << "xyle conserved residues:\n";
//	for (size_t i = 0; i < xyle.size(); ++i) {
//		cout << xyle[i] << '\n';
//	}
//	cout << '\n';
	
	//make pairs
	vector<pair<int, int>> conserved_res;
	cout << "hxt36 xyle" << '\n';
	for (size_t i = 0; i < hxt36.size(); ++i) {
		conserved_res.push_back(make_pair(hxt36[i], xyle[i]));
//		conserved_res[i].first = hxt36[i];
//		conserved_res[i].second = xyle[i];
		cout << conserved_res[i].first << "    " << conserved_res[i].second << '\n';
	}
	
	return 0;
}

File: conserved_resides/conserve.py
# Created by Rohith Pentaparthy on 8/3/17.
# Copyright © 2017 Rohith. All rights reserved.
# This program reads in 1 file that contains color data for multiple proteins, 
# and outputs a file that contains only conserved residues

# 1. Read input
filename = input('Enter the file name with its extension: ')
start_res1 = input('Enter the 1st protein\'s starting residue: ')
start_res2 = input('Enter the 2nd protein\'s starting residue: ')
f = open(filename, 'r')


# 2. Split input into 2 strings, remove '\n' and first word
two_list = []
for word in f.readlines():
    two_list.append(word)
a = two_list[0]
a = a[1:len(a)-1]
a = a.split(' ', 1)[1]
b = two_list[1]
b = b[1:len(b)-1]
b = b.split(' ', 1)[1]


# 3. Get the conserved residues for both proteins
protein1 = []
res_num = int(start_res1)
for i in a.split():
    if (i == '1.0'):
        protein1.append(res_num)
    res_num = res_num + 1
    
protein2 = []
res_num = int(start_res2)
for i in b.split():
    if (i == '1.0'):
        protein2.append(res_num)
    res_num = res_num + 1


# 4. Print out the conserved residue pairs in to a new file
outfile = open('res_conserved.txt', 'w')
outfile.write('hxt36  xyle\n')
for i in range(0, len(protein1)):
    res_pair =  str(str(protein1[i]) + "   " + str(protein2[i]))
    outfile.write(res_pair)
    outfile.write('\n')
num_pairs = str('Number of conserved residue pairs: ' + str(len(protein1)))
outfile.write(num_pairs)


# 5. Close everything
outfile.close()
f.close()